services/src/main/java/org/keycloak/services/resources/LoginActionsService.java

```diff
@@ -1021,5 +1029,14 @@ public void ignore() {
 
         return BrowserHistoryHelper.getInstance().saveResponseAndRedirect(session, authSession, response, true, request);
     }
+    
+    private boolean isCancelAppInitiatedAction(AuthenticationSessionModel authSession, RequiredActionContextResult context) {
+        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
+        
+        boolean userRequestedCancelAIA = formData.getFirst(CANCEL_AIA) != null;
+        boolean isAIARequest = authSession.getClientNote(IS_AIA_REQUEST) != null;
+        
+        return isAIARequest && userRequestedCancelAIA;
```

Comment:
I thought I handled that elsewhere, but now I'm thinking I need to take another look.

Reply:
Now I remember.  It's what you warned me about when we were waiting for the bus in Brno.

This scenario can't happen because AIA is always processed last.  Also, I made sure that submitting cancel doesn't remove a required action.

But I'll double-check to make sure it's all correct.